### ISSUE TITLE
Make it possible to cache when searching by city

### ISSUE THREAD

**Issue by sanders41** [sanders41/weather-command#247]
Currently the data is only cached when searching by zip because the url for city uses longitude and latitude when making the call for weather information while zip uses the zip code. I'm thinking it may work to add the url to the cache and compare that to the built url to see if cache can be used.

**Comment by Himanshuch1708** [comment]
Yes, it is possible to add the URL to the cache and compare it with the built URL to determine if the cache can be used. This approach is commonly used in web development to optimize the performance of web applications.

To implement this approach, you can create a cache key based on the URL and store the weather data in the cache using this key. When a new request is received, you can build the URL and compare it with the cache key. If the cache key matches, you can retrieve the weather data from the cache and return it to the user.

```sh
import requests
import hashlib

def get_weather_data(city=None, zip=None):
    url = build_url(city=city, zip=zip)
    cache_key = hashlib.sha256(url.encode('utf-8')).hexdigest()
    weather_data = cache.get(cache_key)

    if weather_data is None:
        response = requests.get(url)
        weather_data = response.json()
        cache.set(cache_key, weather_data)

    return weather_data
```
In this example, build_url() is a function that constructs the API URL based on the provided city or zip code. cache is an object that provides caching functionality, such as the redis library.

By using this approach, you can avoid making unnecessary API calls and improve the performance of your application.

**Comment by sanders41** [comment]
The location url building and caching happens in [this](https://github.com/sanders41/weather-command/blob/ddd2a0e1f3ec304715a0d2b8a6f18b8753855b70/weather_command/_location.py#L25) function and the weather caches are [here](https://github.com/sanders41/weather-command/blob/ddd2a0e1f3ec304715a0d2b8a6f18b8753855b70/weather_command/_weather.py#L20), [here](https://github.com/sanders41/weather-command/blob/ddd2a0e1f3ec304715a0d2b8a6f18b8753855b70/weather_command/_weather.py#L41), [here](https://github.com/sanders41/weather-command/blob/0b6cfb1e2d44910236501f5b9920c24350a4d03f/weather_command/_builder.py#L443), and [here](https://github.com/sanders41/weather-command/blob/0b6cfb1e2d44910236501f5b9920c24350a4d03f/weather_command/_builder.py#L488). The caching is managed by the class in [this](https://github.com/sanders41/weather-command/blob/main/weather_command/_cache.py) file.

There are functions building the urls already. What may work is to pass those url's into the class and have it generate the hash for a cache key. I haven't dug too far into it yet, but my first inclination is it keep the zip caching as it is now because it will be slightly faster since it doesn't need to build the urls first. If that gets too complicated we could go to using url for everything. The `how` parameter will either have a value of `city` or `zip` so you will always know what type the `city_zip` is from that.

Do you want to see if you can get it working?

**Comment by Himanshuch1708** [comment]
I am currently reviewing and working on your issues. If I am successful in resolving them, I will provide you with an update.

**Comment by sanders41** [comment]
> I am currently reviewing and working on your issues. If I am successful in resolving them, I will provide you with an update.

Sounds good. If you have any questions along the way feel free to ask.
